### PR TITLE
[FLOC-3473] Try to reduce control service load

### PR DESCRIPTION
**Pull request by itamarst**
Instead of a cache that is tied to a particular context, use a LRU cache so delayed updates also benefit from the cache.

### CODE REVIEW

File: flocker/control/_protocol.py [hunk 1]
```diff
@@ -123,46 +124,27 @@ def fromBox(self, name, strings, objects, proto):
         self.another_argument.fromBox(name, strings, objects, proto)
 
 
-class CachingEncoder(object):
-    """
-    Cache results of ``wire_encode`` and re-use them, relying on the fact
-    we're encoding immutable objects.
+# The configuration and state can get pretty big, so don't want too many:
+_wire_encode_cache = LRUCache(50)
```
Comment (dwgebler) [inline flocker/control/_protocol.py]: What's defined as pretty big and too many? I guess we can watch this value over time and modify the cache size if necessary.

File: flocker/control/_protocol.py [hunk 2]
```diff
@@ -123,46 +124,27 @@ def fromBox(self, name, strings, objects, proto):
...
-        self._cache = None
-
-_caching_encoder = CachingEncoder()
+    :param obj: Object to encode.
+    :return: Resulting ``bytes``.
+    """
+    result = _wire_encode_cache.get(obj)
+    if result is None:
+        result = wire_encode(obj)
+        _wire_encode_cache.put(obj, result)
```
Comment (dwgebler) [inline flocker/control/_protocol.py]: As you've noted, this relies on the cached objects' immutability. That makes this look a little bit flaky, but actually we already rely on the type of objects we serialize being immutable in our existing context cache, which I think are mostly ``Deployment`` and maybe one or two other similar models anyway.

**Comment by dwgebler** [comment]
Thanks @itamarst I think this is good to merge when you can get a green build. Cause and solution in the issue make sense, though it'd be interesting to see some metrics on how much of a performance benefit this delivers.

There was a minor merge conflict with ``master`` I have fixed.

**Comment by dwgebler** [comment]
Also you probably have to do something to fix the homebrew recipe builder's failure where it's complaining about not finding `repoze.lru` - I don't know why this happens, it's already a dependency we use elsewhere and specify in ``requirements.txt``.

**Comment by itamarst** [comment]
Weird, it does seem to have installed repoze.lru...

**Pull request merged**